lib/websocket.go

```diff
@@ -58,14 +103,140 @@ func (h *wsHandler) wsMessageHandler(_ context.Context, e event.Event) error {
 		"data":      e.Payload,
 	}
 
-	log.Debugf("sending event %q to %d websocket conns", e.Type, len(h.conns))
-	for k, c := range h.conns {
-		go func(k int, c *websocket.Conn) {
-			err := wsjson.Write(ctx, c, evt)
-			if err != nil {
-				log.Errorf("connection %d: wsjson write error: %s", k, err)
-			}
-		}(k, c)
+	profileIDString := e.ProfileID
+	if profileIDString == "" {
+		log.Debugf("Event with SessionID %q has no scope. Not sending event over websocket.", e.SessionID)
+		return nil
+	}
+	connID, ok := h.subscriptions[profileIDString]
+	if !ok {
+		return fmt.Errorf("no websocket connection ID found for profile %q", profileIDString)
+	}
+	c, ok := h.conns[connID]
+	if !ok {
+		h.UnsubscribeConn(profileIDString)
+		return fmt.Errorf("no websocket connection found for connection ID %q, profile %q", connID, profileIDString)
+	}
+	log.Debugf("sending event %q to websocket conns %q", e.Type, profileIDString)
+	err := wsjson.Write(ctx, c.conn, evt)
+	if err != nil {
+		log.Errorf("connection %q: wsjson write error: %s", profileIDString, err)
+	}
+	return nil
+}
+
+// SubscribeConn authenticates the given token and
+func (h *wsHandler) SubscribeConn(connID, tokenString string) error {
+	ctx := context.TODO()
+	tok, err := token.ParseAuthToken(ctx, tokenString, h.tokenSource)
+	if err != nil {
+		return err
 	}
+
+	claims, ok := tok.Claims.(*token.Claims)
+	if !ok || claims.Subject == "" {
+		return fmt.Errorf("cannot get profile.ID from token")
+	}
+	// TODO(b5): at this point we have a valid signature of a profileID string
+	// but no proof that this profile is owned by the key that signed the
+	// token. We either need ProfileID == KeyID, or we need a UCAN. we need to
+	// check for those, ideally in a method within the profile package that
+	// abstracts over profile & key agreement
```

Comment:
this TODO is pulled from `inst.activeProfile`, where we do the same mining for the profileID using the token. Should probably be pulled into its own method in... token? it would return a (string, error), erroring if the token is invalid or if the profileID does not exist on the token